GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs

```diff
@@ -224,106 +232,140 @@ internal void AddColumn(ColumnProvider columnProvider)
         {
             _lastRepaint.Restart();
 
             DebugHelpers.Assert(_rowHeight != 0, "_rowHeight != 0");
 
-            GitRevision? revision = GetRevision(e.RowIndex);
-
             if (e.RowIndex < 0 ||
                 e.RowIndex >= RowCount ||
-                !e.State.HasFlag(DataGridViewElementStates.Visible) ||
-                revision is null)
+                !e.State.HasFlag(DataGridViewElementStates.Visible))
             {
                 return;
             }
 
-            Brush backBrush = GetBackground(e.State, e.RowIndex, revision);
-            e.Graphics!.FillRectangle(backBrush, e.CellBounds);
+            GitRevision? revision = GetRevision(e.RowIndex);
+            bool isSelected = e.State.HasFlag(DataGridViewElementStates.Selected);
+            if (revision != _revision || _isSelected != isSelected)
+            {
+                _revision = revision;
+                _isSelected = isSelected;
+                if (_revision is null)
+                {
+                    _cellStyle = null;
 
-            if (Columns[e.ColumnIndex].Tag is ColumnProvider provider)
+                    return;
+                }
+
+                Brush backBrush = GetBackground(isSelected, e.RowIndex, _revision);
+                bool isNonRelativeGray = _revisionGraphDrawNonRelativesTextGray && !RowIsRelative(e.RowIndex);
+                Color foreColor = GetForeground(isSelected, isNonRelativeGray);
+                Color commitBodyForeColor = GetCommitBodyForeground(isSelected, isNonRelativeGray);
+                _cellStyle = new(backBrush, foreColor, commitBodyForeColor, _normalFont, _boldFont, _monospaceFont);
+            }
+
+            if (_cellStyle is null)
             {
-                Color foreColor = GetForeground(e.State, e.RowIndex);
-                Color commitBodyForeColor = GetCommitBodyForeground(e.State, e.RowIndex);
-                CellStyle cellStyle = new(backBrush, foreColor, commitBodyForeColor, _normalFont, _boldFont, _monospaceFont);
+                return;
+            }
+
+            e.Graphics!.FillRectangle(_cellStyle.Value.BackBrush, e.CellBounds);
+
+            if (_revision is null)
+            {
+                return;
+            }
 
-                provider.OnCellPainting(e, revision, _rowHeight, cellStyle);
+            if (Columns[e.ColumnIndex].Tag is ColumnProvider provider)
+            {
+                provider.OnCellPainting(e, _revision, _rowHeight, _cellStyle.Value);
             }
 
             if (!e.Handled)
             {
                 e.Handled = true;
-                _forceRefresh = true;
-                UpdateVisibleRowRange();
+
+                // Ensure triggering is done only one time per row
+                if (e.ColumnIndex == _unhiddableColumnMessageIndex)
+                {
+                    _forceRefresh = true;
+                    UpdateVisibleRowRange();
+                }
```

Comment:
Good finding!
I had already started to rewrite `UpdateVisibleRowRange` including the graph rendering. And I am going to continue after merging this PR.
But I think here, we should rather let the other column providers properly set `e.Handled = true`. As far as I remember, the graph is the only one which really needs to wait for additional data. The others should be happy with having a non-null revision.

Reply:
I would even set `e.Handled = true` by default before calling `provider.OnCellPainting`. If a providers needs an async update, it shall reset the `e.Handled` to `false`.
The longer I think about it, the less this `_forceRefresh = true` makes sense to me here. There is no reason to invalidate the entire current graph cache.

Reply:
regarding `e.Handled`: f33646202667a9fc59c57a462f751e65755e3273